### ISSUE TITLE
needless_lifetimes: false positive with impl Trait

### ISSUE THREAD

**Issue by japaric** [Manishearth/rust-clippy#1306]
This code compiles:

``` rust
#![feature(conservative_impl_trait)]

trait Foo {}

struct Bar {}

struct Baz<'a> {
    bar: &'a Bar,
}

impl<'a> Foo for Baz<'a> {}

impl Bar {
    fn baz<'a>(&'a self) -> impl Foo + 'a {
        Baz { bar: self }
    }
}

fn main() {}
```

But clippy gives this warning:

```
$ cargo clippy
warning: explicit lifetimes given in parameter types where they could be elided, #[warn(needless_lifetimes)] on by default
  --> src/main.rs:14:5
   |
14 |     fn baz<'a>(&'a self) -> impl Foo + 'a {
   |     ^
   |
   = help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#needless_lifetimes
```

If you apply the suggestion:

``` diff
 impl<'a> Foo for Baz<'a> {}

 impl Bar {
-    fn baz<'a>(&'a self) -> impl Foo + 'a {
+    fn baz(&self) -> impl Foo {
         Baz { bar: self }
     }
 }
```

you end with code that doesn't compile:

```
$ cargo clippy
error[E0564]: only named lifetimes are allowed in `impl Trait`, but `` was found in the type `Baz<'_>`
  --> src/main.rs:14:22
   |
14 |     fn baz(&self) -> impl Foo {
   |                      ^^^^^^^^

error: aborting due to previous error
```

I don't know if these lifetimes will become elidable in the future but, in this case, clippy should probably not raise a warning for now.
